### ISSUE TITLE
DioError [DioErrorType.DEFAULT]: type 'Future<dynamic>' is not a subtype of type 'FutureOr<Response<dynamic>>'

### ISSUE THREAD

**Issue by yanhao98** [flutterchina/dio#817]
### New Issue Checklist

- [ ] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | Dart                                                  |      |
| Platform Version               | e.g.  1.5.0 / 12.0 / 9.0                              |      |
| Dio Version                    | dio: ^3.0.9                                           |      |
| Android Studio / Xcode Version | e.g. Android Studio 3.3.2 / Xcode 10.2.1              |      |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |      |
| Repro with our demo prj        | e.g. does it happen with our demo project?            |      |
| Demo project link              | e.g. link to a demo project that highlights the issue |      |

### Issue Description and Steps

我的需求是在 Response 拦截器中判断如果code != 0.抛出一个DIoError。

然后在 Error 拦截器中refreshToken。

但是在我的这份代码中，lock错误拦截器。就会抛出

> DioError [DioErrorType.DEFAULT]: type 'Future<dynamic>' is not a subtype of type 'FutureOr<Response<dynamic>>'

我不知道为什么，如果是我的代码有问题，还请指出。谢谢。

这是我的代码，上传在网盘

下载地址：

> 链接: https://pan.baidu.com/s/1OaPYfIMUHtceEP99NjPS9A 提取码: p24x

![image-20200603162958408](https://i.loli.net/2020/06/03/Z9TeiJ48tAqW6lk.png) 

![image-20200603161702617](https://i.loli.net/2020/06/03/FxdQUCEsIZSlm1f.png)

**Comment by yanhao98** [comment]
有时会有，有时不会报错。

**Comment by JDNew** [comment]
> 有时会有，有时不会报错。

我也遇到类似的情况，我的错误是 "DioError [DioErrorType.DEFAULT]: type 'Future<DioError>' is not a subtype of type 'FutureOr<RequestOptions>'"

**Comment by JDNew** [comment]
Me too .

**Comment by leechan** [comment]
Same problem, 3.0.9

**Comment by leechan** [comment]
@yanhao98 有找到解决方法吗？咱们几乎完全相同的使用场景
